pymc3/distributions/distribution.py

```diff
@@ -141,40 +245,116 @@ def __new__(cls, name, *args, **kwargs):
                 "for a standalone distribution."
             )
 
-        rng = kwargs.pop("rng", None)
-
-        if rng is None:
-            rng = model.default_rng
-
         if not isinstance(name, string_types):
             raise TypeError(f"Name needs to be a string but got: {name}")
 
-        data = kwargs.pop("observed", None)
-
-        total_size = kwargs.pop("total_size", None)
-
-        dims = kwargs.pop("dims", None)
-
-        if "shape" in kwargs:
-            raise DeprecationWarning("The `shape` keyword is deprecated; use `size`.")
+        if rng is None:
+            rng = model.default_rng
 
-        transform = kwargs.pop("transform", UNSET)
+        # Create the RV without specifying testval, because the testval may
+        # have a shape that only matches after replicating with a size implied
+        # by dims (see below).
+        rv_out = cls.dist(*args, rng=rng, testval=None, **kwargs)
+
+        # `dims` are only available with this API, because `.dist()` can be used
+        # without a modelcontext and dims are not yet tracked at the Aesara level.
+        if dims is not None:
+            # Infer size from dims (and coords)
+            size = tuple(len(model.coords[dname]) for dname in dims if dname != Ellipsis)
+            if Ellipsis in dims:
+                # Auto-complete the dims tuple with default dim names
+                S = len(size)
+                dims = tuple(dims[:S]) + tuple(f"{name}_dim_{d}" for d in range(S, S + rv_out.ndim))
+                for dname in dims:
+                    if not dname in model.coords:
+                        model.coords[dname] = None
```

Comment:
We shouldn't care about that. If the dimension name is not in `coords` that dimension will have default coordinate values (ArviZ) or will have no coordinate values (pure xarray). Both do different things but use the same "convention" in that default/nonexistent coord values are indicated by the dim name not being a key in the coords dict. 

It's not an issue at all if not all dims have coord values defined in `coords` whereas using `None` will indicate that there are coord values (the key is present in the coords dict) and both ArviZ and xarray will try to use `None` as coordinate values and error out.

Reply:
So the `dims` can be `("country", "town", None, None)`, but for the `None` elements the `model.coords` should remain unchanged?

Asking because we probably want a tuple of the correct length in the `model.RV_dims` dictionary: https://github.com/pymc-devs/pymc3/pull/4625/files/4c60766db3ace7765c1f5b0696cc04b70b608b25#diff-6a4ca81dc9f970173ecf8842d1f63fa74af138f33a5d88bbc6620e2e2e493a01R284

Reply:
For ArviZ to handle that automatically, it should be `("country", "town")`. If we want to have `dims` with the same length as the variable we can remove the `None`s in `to_inferencedata`. 

I think we should go with the first option though, with no `None`s, precisely because we will use `...` to indicate that the number of dimensions can change right? Then there is no "same number of dimensions as the actual rv"

Reply:
To the best of my knowledge the _number of dimensions_ cannot change.

Their length should be flexible:
```python
x = aesara.shared(np.ones(5))
x.set_value(np.ones(10))
```

IIRC I tried to include such a re-sizing of a shared variable in the test case for shape asserts, but it didn't work yet.

Reply:
That's correct; the number of dimensions cannot change.